### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by lucaasdfg**
third milestone

### CODE REVIEW

**Comment by lucaasdfg** [comment]
the stickler error is because it's already in development the file, once we push the error will be gone

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Changes required

### Hi team! :wave: 

You have made some progress, but there are changes we need to make:

1. There is no lib folder, in this branch, here is the reason:
```Game_logic branch is 6 commits behind development. ```  Research on how to fix it.
2. The classes you implemented in main.rb, should be in the lib folder; then the files will imported from lib  folder to main.rb.
3. Executing the main.rb file, and following the instruction, when I got to "mark your choice", all the choice I choose were invalid. While fixing this issue, I would suggest, the choice be stated explicitly.

Peek 2019-11-09 07-28


When you are done implementing this mile stone, this is how the game should behave:

##### Behaviour

* [x] The application does not not allow a player to move into an occupied position
* [x] The application does not exit due to invalid input - it informs player about it and asks for their move again
* [x] The application saves players move and displays status of board after that.
* [ ] The application can determine if move is a winning move
  * [ ] The application can detect a sequence of three horizontal elements that are equal
  * [ ] The application can detect a sequence of three vertical elements that are equal
  * [ ] The application can detect a sequence of three equal elements in a diagonal from left and right to right
* [ ] The application can tell if there was a draw after players move
* [ ] The application can change active player

- [ ] Also, fix the linter error.

When you are done implementing the changes, submit another form for a code review.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Comment by lucaasdfg** [comment]
the linter error will go away after merge

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required

The changes suggested by @uche-inyama  are yet to be implemented.

Let's also address this:
* We don't have stickler configuration files in place for this repo. Visit [this](https://github.com/microverseinc/linters-config/tree/master/ruby) and include all the config file required for the stickler to work.
* All the game logic code should reside in the `lib` directory. Let's remove all the game logic classes from `bin/main.rb`

**After implementing the changes, submit another request for review.**

_Happy Coding :computer:  
Reviewed by: **Simon Wathigo**_

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required  

Let's address the following issues:
* The game does not explicitly indicate the winner after a wining move. See this :point_down: 
ohh.

**[OPTIONAL]**
I think we can make the board look better while playing.
hjewf


**After implementing the changes, submit another request for review.**

_Happy Coding :computer:  
Reviewed by: **Simon Wathigo**_

File: lib/gamelogic.rb
Comment (stratospherique) [inline lib/gamelogic.rb]: - Taking the case where there is a winner at the last move, you code prevent the game from declaring a winner because you are overwriting the result value after you exit the each loop. :pray: fix it.

File: lib/gamelogic.rb
Comment (stratospherique) [inline lib/gamelogic.rb]: * [ ] No `puts/print` and `get.chomp` outside `main.rb` file.
  - All code responsible for UI interactions should go inside `main.rb`, sorry for not mentioning this in previous reviews, but it's a requirement.
  - These methods can go inside `main.rb` after doing the necessary changes :nerd_face:

File: lib/gamelogic.rb
Comment (stratospherique) [inline lib/gamelogic.rb]: - Same note about `puts/print`

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great progress Team :clap: Almost done. However, to get approval for this milestone, the changes below needs to be implemented :arrow_down: 

- **[REQUIRED]** Check code comments about the winner declaration issue and Code structuring requirements.

- **[SUGGESTION]** When switching turns, you should display each player name.

### That's it :relaxed: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
_**Ahmed Mahfoudh**_

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: `attr_accessor` is not allowed.
Please, choose between `attr_reader` or `attr_writer`

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: As say up, `attr_accessor` is not allowed

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: Instead of using string `'X'` and `'O'`, please use symbol, I mean `:X` and `:O`

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: Make it clear please.
You could write it in many lines..

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: Use symbol here too.
Use `:VICTORY` or `:DRAW`

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: I don't see it in the Board class. Not good usage of OOP.

File: lib/gamelogic.rb
Comment (maelfosso) [inline lib/gamelogic.rb]: As say up, use symbol

File: bin/main.rb
Comment (maelfosso) [inline bin/main.rb]: From an OOP point of view, this is not good.
The Board is an integral part of the game. It cannot exists outside the game, it cannot exist if the game does not exist.

From that statement, If you instantiate the Game, the game comes to the world with the Board (and the players) and if you destroy the Game, it dies after killing himself the Board (and the players)

That is the **OOP**. All that talk to tell you guys to have the board as an attribute of the game.
```ruby
class Game
   attr_reader :board 
   ....
end
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello @lucaasdfg @alexawesomecode 

Nice game !!! It was a pleasure to play it :smile: 
However, even if it's OK in the external side (the game is OK - Winner and Draw, Error, etc...) I have some suggestions to improve the internal side (the code)

1. Please, use one file for each class. I mean one for Game, another one for Board

2. Refactor your code such that the **Game** class have the board as an attribute. 

3. Read the comment to know more about the others problems.

Kindly make the changes and submit a new code review request.
Thank you

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: bin/main.rb
Comment (lucaasdfg) [reply bin/main.rb]: thank, I think I get the idea now. but I'm not sure how to fix it. could you give me any advice or a link to documentation or tutorial

File: bin/main.rb
Comment (ebukaume) [inline bin/main.rb:51]: You should not disable code linting  like this

File: lib/game.rb
Comment (ebukaume) [inline lib/game.rb]: This is a very bad practice. LineLength is a very important metric; You should not disable it.

File: lib/players.rb
Comment (ebukaume) [inline lib/players.rb]: Why do you need this `attr_writer`? Do you understand the function? 

Are you expecting to overwrite your players array?

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes Required!

Hi team,

Note that the style guide is there to make us stick to established patterns and not for us to **bypass** them :wink: 

Let's remove those in-line rubocop directives and get our code to pass the checks. you can use multiline expression. It makes your code readable.

**See below for more**

**Kindly submit another code review request after fixing those**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Review by ebukaume — APPROVED** [review]
It is much better now. **You can go ahead and merge this branch.**


Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Pull request merged**
